lib/discordgo/user.go

```diff
@@ -72,13 +73,25 @@ func (u *User) NKeys() int {
 }
 
 // AvatarURL returns a URL to the user's avatar.
-//    size:    The size of the user's avatar as a power of two
-//             if size is an empty string, no size parameter will
-//             be added to the URL.
+//
+//	size:    The size of the user's avatar as a power of two
+//	         if size is an empty string, no size parameter will
+//	         be added to the URL.
 func (u *User) AvatarURL(size string) string {
 	var URL string
 	if u.Avatar == "" {
-		URL = EndpointDefaultUserAvatar(u.Discriminator)
+		// See https://discord.com/developers/docs/reference#image-formatting-cdn-endpoints:
+		//
+		// "For users on the new username system, `index` will be `(user_id >> 22) % 6`.
+		// For users on the legacy username system, `index` will be `discriminator % 5`."
+		var index int
+		if u.Discriminator == "0" {
+			index = int((u.ID >> 22) % 6)
+		} else {
+			discrim, _ := strconv.Atoi(u.Discriminator)
+			index = discrim % 5
+		}
```

Comment:
I considered factoring this bit out into a separate `DefaultAvatarIndex` method — notably, this is what upstream d.go does — but decided against it because the logic's fairly simple and there's no use that I can see for it elsewhere. Happy to change that if the separate method seems worthwhile.